drivers/console/aarch64/console.S

```diff
@@ -3,103 +3,240 @@
 
 	/*
-	 *  The console base is in the data section and not in .bss
-	 *  even though it is zero-init. In particular, this allows
+	 *  The console list pointer is in the data section and not in
+	 *  .bss even though it is zero-init. In particular, this allows
 	 *  the console functions to start using this variable before
 	 *  the runtime memory is initialized for images which do not
 	 *  need to copy the .data section from ROM to RAM.
 	 */
-.section .data.console_base ; .align 3
-	console_base: .quad 0x0
+.section .data.console_list ; .align 3
+	console_list: .quad 0x0
+.section .data.console_state ; .align 0
+	console_state: .byte (1 << CONSOLE_FLAG_BOOT_SHIFT)
 
 	/* -----------------------------------------------
+	 * void console_register(console_info_t *console)
+	 * Function to insert a new console structure into
+	 * the console list. Should usually only be called
+	 * by declare_console expansions.
+	 * In : x0 - address of console structure
+	 * Clobber list: x0, x1, x14
+	 * -----------------------------------------------
+	 */
+func console_register
+	adrp	x14, console_list
+	ldr	x1, [x14, :lo12:console_list]	/* X1 = first struct in list */
+	str	x0, [x14, :lo12:console_list]	/* list head = new console */
+	str	x1, [x0]			/* new console next ptr = X1 */
+	ret
+endfunc console_register
+
+#if !ERROR_DEPRECATED
+	/* Set up compatiblity symbols for platforms using old console API. */
+	declare_console core
+
+	.globl	console_init
+	/* -----------------------------------------------
 	 * int console_init(uintptr_t base_addr,
 	 * unsigned int uart_clk, unsigned int baud_rate)
-	 * Function to initialize the console without a
-	 * C Runtime to print debug information. It saves
-	 * the console base to the data section.
+	 * Deprecated legacy function for initializing a
+	 * single console. Current platforms should call
+	 * console_<driver>_register for the relevant
+	 * driver instead. This needs to clear the console
+	 * list first because some legacy platforms called
+	 * console_init more than once and expect it to
+	 * overwrite previous console state.
 	 * In: x0 - console base address
 	 *     w1 - Uart clock in Hz
 	 *     w2 - Baud rate
-	 * out: return 1 on success else 0 on error
-	 * Clobber list : x1 - x4
+	 * Clobber list : x0 - x4, x14, x15
 	 * -----------------------------------------------
 	 */
 func console_init
-	/* Check the input base address */
-	cbz	x0, init_fail
-	adrp	x3, console_base
-	str	x0, [x3, :lo12:console_base]
-	b	console_core_init
-init_fail:
-	ret
+	adrp	x14, console_list
+	str	xzr, [x14, :lo12:console_list]
+	b	console_core_register
 endfunc console_init
+#endif
+
+	/* -----------------------------------------------
+	 * console_info_t *console_unregister(
+	 *                        console_info_t *console)
+	 * Function to find a specific console in the list
+	 * of currently active consoles and remove it.
+	 * Should usually only be called by
+	 * declare_console expansions.
```

Comment:
> I would like to get away from the forced console_init(baseaddr, clock, baud) function signature, because for some consoles those parameters may not make sense and they'll want different ones instead. That's why every driver can declare their own console_xxx_register() function (e.g. the one in include/drivers/ti/uart/uart_16550.h) with the specific signature they need.

Yes, I have no objection to that. I was just saying it's confusing to have console_register() in console.h when the expectation is that clients call console_xxx_register().

> In fact, now that I think of it it's not really useful to have C declarations for console_register() and console_unregister() anyway. Let me remove those so it's an implementation detail of the assembly code.

Agreed.